contracts/allocators/OnsenAllocatorV2.sol

```diff
@@ -0,0 +1,268 @@
+                }
+            }
+        }
+    }
+
+    function _deactivate(bool panic) internal override {
+        /// Get amounts by LP to unstake and then deallocate (unstake) each token.
+        _deallocateAll();
+        _unStakeSushi();
+
+        /// If Panic transfer LP tokens to the treasury
+        if (panic) {
+            for (uint256 i; i < _tokens.length; i++) {
+                IERC20 token = _tokens[i];
+                token.safeTransfer(treasury, token.balanceOf(address(this)));
+            }
+
+            /// And transfer Sushi tokens (rewards) to treasury
+            IERC20(sushi).safeTransfer(treasury, IERC20(sushi).balanceOf(address(this)));
+        }
+    }
+
+    function _prepareMigration() internal override {
+        /// Get amounts by LP to unstake and then deallocate (unstake) each token.
+        _deallocateAll();
+        _unStakeSushi();
+    }
+
+    function amountAllocated(uint256 id) public view override returns (uint256) {
+        /// Find the LP token which amount allocated is requested
+        uint256 index = tokenIds[id];
+        IERC20 LPtoken = _tokens[index];
+        PoolData memory pd = _lpToOnsenId[address(LPtoken)];
+
+        if (pd.poolActive) {
+            UserInfo memory currentUserInfo = IMasterChef(masterChef).userInfo(pd.id, address(this));
+            return currentUserInfo.amount;
+        } else {
+            return 0;
+        }
+    }
+
+    function rewardTokens() public view override returns (IERC20[] memory) {
+        IERC20[] memory rewards = new IERC20[](2);
+        rewards[0] = IERC20(sushi);
+        rewards[1] = IERC20(xSushi);
+        return rewards;
+    }
+
+    function utilityTokens() public view override returns (IERC20[] memory) {
+        IERC20[] memory utility = new IERC20[](1);
+        utility[0] = IERC20(sushi);
+        return utility;
+    }
+
+    function name() external pure override returns (string memory) {
+        return "OnsenAllocator";
+    }
+
+    function setTreasury(address treasuryAddress) external onlyGuardian {
+        treasury = treasuryAddress;
+    }
+
+    function setSushi(address sushiAddress) external onlyGuardian {
+        sushi = sushiAddress;
+    }
+
+    function setXSushi(address xSushiAddress) external onlyGuardian {
+        xSushi = xSushiAddress;
+    }
+
+    function setMasterChefAddress(address masterChefAddress) external onlyGuardian {
+        masterChef = masterChefAddress;
+    }
+
+    /// ========== INTERNAL FUNCTIONS ==========
+
+    /**
+     * @notice stake sushi rewards
+     */
+    function _stakeSushi() internal {
+        uint256 balance = IERC20(sushi).balanceOf(address(this));
+        if (balance > 0) {
+            IERC20(sushi).approve(xSushi, balance);
+            ISushiBar(xSushi).enter(balance); // stake sushi
+        }
+    }
+
+    /**
+     * @notice unStake sushi rewards
+     */
+    function _unStakeSushi() internal {
```

Comment:
fairly unimportant but if u using camel case the first s is lowercase because unstake is a whole word.

Reply:
Function name updated.